lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -536,7 +536,24 @@ void prepareCacheControlsTranslation(Metadata *MD, Instruction *Inst) {
 
     // Create dummy GEP for SSA copy of the pointer operand. Lets do our best
     // to guess pointee type here, but if we won't - just pointer is also fine,
-    // if necessary TypeScavenger will adjust types and create bitcasts.
+    // if necessary TypeScavenger will adjust types and create bitcasts. If
+    // memory instruction operand is already zero GEP - create nothing and use
+    // the old GEP.
+    SmallVector<Metadata *, 4> MDs;
+    std::vector<Metadata *> OPs = {KindMD, LevelMD, ControlMD};
+    if (auto *const GEP = dyn_cast<GetElementPtrInst>(PtrInstOp)) {
+      if (GEP->hasAllZeroIndices()) {
+        MDs.push_back(MDNode::get(Inst->getContext(), OPs));
+        // If the existing GEP has SPIRV_MD_DECORATIONS metadata - copy it
+        if (auto *OldMD = GEP->getMetadata(SPIRV_MD_DECORATIONS))
+          for (unsigned I = 0, E = OldMD->getNumOperands(); I != E; ++I)
+            if (auto *DecoMD = dyn_cast<MDNode>(OldMD->getOperand(I)))
+              MDs.push_back(DecoMD);
+        MDNode *MDList = MDNode::get(Inst->getContext(), MDs);
+        GEP->setMetadata(SPIRV_MD_DECORATIONS, MDList);
+        return;
+      }
+    }
     IRBuilder Builder(Inst);
     Type *GEPTy = PtrInstOp->getType();
```

Comment:
```suggestion
    Type *GEPTy = Builder.getInt8Ty();
```

Not having `i8` here leads to the type scavenger to introduce a `bitcast` in some cases and will mess up with the cache control MD. If we process:
```llvm
source_filename = "LLVMDialectModule"                                                                                                                                                                                                                      
target triple = "spir64"                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                           
; Function Attrs: nounwind                                                                                                                                                                                                                                 
declare spir_func void @_Z40intel_sub_group_2d_block_read_8b_8r32x2cPU3AS1viiiDv2_iPt(ptr addrspace(1) nonnull readonly, i32, i32, i32, <2 x i32>, ptr nonnull writeonly) #0                                                                               
                                                                                                                                                                                                                                                           
define spir_func void @triton_gen.2Dblockload(ptr addrspace(1) %0, ptr addrspace(1) %1, i32 %2, i32 %3, i32 %4, i32 %5, i32 %6, ptr %7) {                                                                                                                  
entry:                                                                                                                                                                                                                                                     
  %8 = insertelement <2 x i32> undef, i32 %5, i32 0                                                                                                                                                                                                        
  %9 = insertelement <2 x i32> %8, i32 %6, i32 1                                                                                                                                                                                                           
  call spir_func void @_Z40intel_sub_group_2d_block_read_8b_8r32x2cPU3AS1viiiDv2_iPt(ptr addrspace(1) %0, i32 %2, i32 %3, i32 %4, <2 x i32> %9, ptr %7), !spirv.DecorationCacheControlINTEL !1                                                             
  ret void                                                                                                                                                                                                                                                 
}                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                           
attributes #0 = { nounwind }                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                           
!llvm.module.flags = !{!0}                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                           
!0 = !{i32 2, !"Debug Info Version", i32 3}                                                                                                                                                                                                                
!1 = !{!2, !3}                                                                                                                                                                                                                                             
!2 = !{i32 6442, i32 0, i32 3, i32 0}                                                                                                                                                                                                                      
!3 = !{i32 6442, i32 1, i32 1, i32 0}
```

using this version with:

```
llvm-as input.ll -o input.bc
llvm-spirv input.bc -o - --spirv-ext=+SPV_INTEL_cache_controls --spirv-text
```

We get:


```
5 Decorate 33 CacheControlLoadINTEL 1 1
5 Decorate 33 CacheControlLoadINTEL 0 3
...
5 Function 2 19 0 18
3 FunctionParameter 17 20
3 FunctionParameter 4 21
3 FunctionParameter 5 22
3 FunctionParameter 5 23
3 FunctionParameter 5 24
3 FunctionParameter 5 25
3 FunctionParameter 5 26
3 FunctionParameter 8 27

2 Label 28
6 CompositeInsert 6 30 25 29 0
6 CompositeInsert 6 31 26 30 1
5 PtrAccessChain 17 33 20 32
4 Bitcast 4 34 33
10 FunctionCall 2 35 10 34 22 23 24 31 27
1 Return
```

As we can see, the function call receives the `bitcast` output as an input. This is misinterpreted by tools like IGC in my tests, leading to the decorations to be dropped.

After my patch, we get:
```
5 Decorate 32 CacheControlLoadINTEL 1 1
5 Decorate 32CacheControlLoadINTEL 0 3
...
5 Function 2 18 0 17
3 FunctionParameter 4 19
3 FunctionParameter 4 20
3 FunctionParameter 5 21
3 FunctionParameter 5 22
3 FunctionParameter 5 23
3 FunctionParameter 5 24
3 FunctionParameter 5 25
3 FunctionParameter 8 26

2 Label 27
6 CompositeInsert 6 29 24 28 0
6 CompositeInsert 6 30 25 29 1
5 PtrAccessChain 4 32 19 31
10 FunctionCall 2 33 10 32 21 22 23 30 26
1 Return
```

No `bitcast` and cache controls decoration being well interpreted by tools. 

I think this case does not show up in your test as the translator does not have a notion of the function you're using and its typing. This may be true for other similar builtin functions.

Having this as a test case might be handy.

Reply:
Thanks for the suggestion and explanation, applied

Reply:
I left existing tests as bitcast is being introduced in one of them